tests/test_class.py

```diff
@@ -0,0 +1,84 @@
+
+cocci_patches_path = Path.cwd() / "cocci"
+input_files_path = Path.cwd() / "tests/data/inputs"
+cocci_outputs_path = Path.cwd() / "cocci_outputs"
+expected_outputs_path = Path.cwd() / "tests/data/expected_outputs"
+
+class Test(unittest.TestCase):
+    def test_cocci_patch_complete(self):
+        """
+        Checks if folder contains all of the patches.
+        """
+        files_in_folder = [file.name for file in cocci_patches_path.iterdir() if file.is_file()]
+        all_files_present = all(file in files_in_folder for file in patch_filenames)
+        if not all_files_present:
+            print(f"\033[91m \nSome Coccinelle patch files are missing.\033[0m")
+        self.assertTrue(all_files_present)
+
+    def test_c_inputs_complete(self):
+        """
+        Checks if all source files are present.
+        """
+        files_in_folder = [file.name for file in input_files_path.iterdir() if file.is_file()]
+        all_files_present = all(file in files_in_folder for file in input_filenames)
+        if not all_files_present:
+            print(f"\033[91m \nSome C source files missing.\033[0m")
+        self.assertTrue(all_files_present)
+
+    def test_make_comparison(self):
+        """
+        Test that Coccinelle produced atoms matches the expected 
+        """
+        failed_cases = []
+
+        for item in expected_outputs_path.glob("*.csv"):
+
+
+            try:
+                output = pd.read_csv(Path(cocci_outputs_path, item.name), header=None)
+                output = output.sort_values(by=output.columns[2])
+                # output.to_csv()
+            except:
+                print(f"{item} not found in cocci_output")
+                continue
+
+            expected = pd.read_csv(Path(expected_outputs_path, item.name), header=None)
+
+            line_numbers_output = output.iloc[:,2]
+            line_numbers_expected = expected.iloc[:,2]
+            
+            missing_in_output = line_numbers_expected[~line_numbers_expected.isin(line_numbers_output)].tolist()
+            missing_in_expected = line_numbers_output[~line_numbers_output.isin(line_numbers_expected)].tolist()
+
+            # print(f"Line numbers in expected but not in output:\n{missing_in_output}")
+            # print(f"Line numbers in output but not in expected:\n{missing_in_expected}")
+                
+            with self.subTest(atom=item.name):
+                try:
+                    self.assertTrue(not len(missing_in_output) and not len(missing_in_expected))
+                except AssertionError:
+                    failed_cases.append(item.name)
+                    self.fail("{item.name} failed")
+                # self.assertTrue(not len(missing_in_output) and not len(missing_in_expected))
+                                                                       
+        if failed_cases:
```

Comment:
I don't have much experience with `unittest` seeing that I tend to use `pytest`, but it seems odd to me that it's necessary to keep track of the failed tests manually. Are there any features of the framework that automate some of this work?

Reply:
unittest will printput all the failed cases and allows for customization such as printing the atom name of each failure. However, I manually created an array to append the atoms just to give a summary of what failed.